### ISSUE TITLE
"Unneeded return statement" in macro

### ISSUE THREAD

**Issue by anka-213** [Manishearth/rust-clippy#1271]
When using the following macro:

```
macro_rules! derr(
    ($($arg:tt)*) => (return Err(Decode(format!($($arg)*))))
);
```

The warning "unneeded return statement" is triggered for every usage of the macro where return would be unnecessary:

```
warning: unneeded return statement, #[warn(needless_return)] on by default
   --> src/dopt.rs:645:23
    |
645 |     ($($arg:tt)*) => (return Err(Decode(format!($($arg)*))))
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
766 |             _ => derr!("Could not decode '{}' into char for '{}'.", vstr, k),
    |                  ----------------------------------------------------------- in this macro invocation
    |
help: remove `return` as shown:
    |     ($($arg:tt)*) => (Err(Decode(format!($($arg)*))))
    = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_return
```

However, the point of the macro is to be able to use it in all positions, so the warning feels a bit redundant. And more importantly, the suggested fix changes the behavior of code in other places. Best case scenario would be a type error, worst case scenario would be silently different behavior.

I could use `#[allow(needless_return)]`, but then I would have to write it at every usage of the macro. Writing it above the macro definition was not enough.

**Comment by daboross** [comment]
Just thought I'd mention I ran into this as well.

Test case on playpen similar to my setup I made before realizing there was an existing issue: http://play.integer32.com/?gist=b8d662faa934cadf292e48006d7ca268&version=undefined.

**Comment by llogiq** [comment]
This is just one more long that could do with a macro check.

This makes a good first PR. If anyone here is interested, please respond on this issue.
